Command Timeout Progress Bar

panzarino opened the pull request:
<!-- 
Thanks for contributing!
Read our contribution guidelines here: 
https://github.com/cypress-io/cypress/blob/develop/CONTRIBUTING.md 
-->

<!-- Example: "Closes #1234" -->

- Closes #7642

### User facing changelog

Adds an animated progress bar indicating how close the running command is to timing out

### Additional details

In order to implement these changes, two new fields were added to every `command` log object: `timeout` and `wallClockStartedAt`. These two fields are needed in order to know what the timeout is for a specific command (each command has different defaults and the ability to change the timeout through `options`), as well as how long the command has been running for when the progress bar is first displayed (displayed at different times based on number of tests in the file and arbitrary user clicks). 

### How has the user experience changed?

Here is an example of the progress bar when a command times out:

![ezgif-6-ed8f7d0ed733](https://user-images.githubusercontent.com/7033952/84438187-1a575880-ac04-11ea-8468-6ea37d820250.gif)

### PR Tasks

<!-- 
These tasks must be completed before a PR is merged.
Delete tasks if they are not applicable. 
-->

- [ ] Have tests been added/updated?
- [ ] Has the original issue been tagged with a release in ZenHub? <!-- (internal team only)-->
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? <!-- Link to PR here -->
- [ ] Have API changes been updated in the [`type definitions`](../cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?

panzarino committed:
Timeout animation

panzarino committed:
Add timeout to log object

panzarino committed:
Merge branch 'develop' into issue-7642

panzarino committed:
Set timeout to log when explicitly overwritten

panzarino committed:
Merge branch 'develop' into issue-7642

panzarino committed:
No need to display if no timeout

panzarino committed:
Throw timeouts everywhere!
I'll clean it up later

panzarino committed:
Correctly set timeout indicator regardless of when test is opened

panzarino committed:
Slightly modify that

panzarino committed:
Update timeout for commands

panzarino committed:
Update progress bar for consistent progress

panzarino committed:
Merge branch 'develop' into issue-7642

panzarino committed:
Fix issue with flashing full progress bar after complete

panzarino committed:
Add test for progress

panzarino committed:
Merge branch 'develop' into issue-7642

JessicaSachs commented on packages/reporter/src/commands/command.tsx:
```diff
@@ -198,6 +199,7 @@ class Command extends Component<Props> {
             </span>
           </div>
         </FlashOnClick>
+        { model.state === 'pending' && !!timeRemaining && <div className='command-progress'><span style={{ animationDuration: `${timeRemaining}ms`, width: `${timeRemaining / model.timeout * 100}%` }} /></div> }
```
Can we pull this out into something?

A function at a minimum but also appropriate would be a component since its state and styles are pretty self-contained.

JessicaSachs commented on packages/reporter/src/commands/command.tsx:
```diff
@@ -134,6 +134,7 @@ class Command extends Component<Props> {
   render () {
     const { model, aliasesWithDuplicates } = this.props
     const message = model.displayMessage
+    const timeRemaining = model.timeout ? model.timeout - ((new Date()).getTime() - model.wallClockStartedAt.getTime()) : 0
```
This is a bit hard to read. I’d usually work with `Date.now` or the `performance.now` results rather than the Date API.

JessicaSachs reviewed (changes requested):
Threading options.timeout throughout the flow seems okay, but I’d like someone else to comment on this.

The UI code could be a bit cleaner and I left suggestions. The one I really care about is isolating the logic in the template into some sort of function or component. The Date stuff is a nitpick and just a suggestion to clean up the code and let you know there’s a diff way to spell it.

panzarino committed:
Move progress bar to function

panzarino committed:
Merge branch 'develop' into issue-7642

panzarino committed:
Update unit test

jennifer-shehane reviewed (changes requested):
This is really cool btw when you use it. 😄 

1. So, it creates this 'jitter' effect due to the height of the command being 2px higher and then settling into the natural height of 2px less. 

	Typically you'd want to negate this by always having the height of the command consistent whether there's a progress bar or not.

	![](http://g.recordit.co/xWoPuZIC1N.gif)

2. The animations still play when using `.debug()`. Not sure how we want to handle this.

	![](http://g.recordit.co/hHpQpPRuhE.gif)

3. I don't think progress bars are meant to be logged for just standard XHR requests / events that are not actual commands from Cypress. There is no 'timeout' for these completing, they respond when they respond.

	![](http://g.recordit.co/RJ8jHZPOHp.gif)

4. This is what I meant when speaking of perf the other day. Some of the team said 'there will only be one progress bar at a time', but if you do something like the example below https://github.com/cypress-io/cypress/issues/6967, it gets overwhelmed drawing 200+ wrap/invokes/asserts in the Command Log, all with progress bars. 

	There is a perf problem in general, so this PR doesn't introduce that!

	With this PR, the example seems to be spending ~4sec more rendering though. And issuing a big warning about layout shift. (Perhaps because we're adding the element in with height and removing it - requiring layout to be recalculated).  

	I'd probably like someone else to verify though because I'm not that great with the performance tools. 

	```js
	it('perf', () => {
	  cy.visit('https://entertainment.cathaypacific.com/catalog?template=movie&parent=Movies')
	  cy.get('.carouselImage > poster-image > figure > picture > source', { timeout: 40000 }).each(source => {
	    cy.wrap(source)
	      .invoke('attr', 'srcset')
	      .should('contain', 'http')
	  })
	})
	```

	![](http://g.recordit.co/HQ3tgXJzXT.gif)

	4.8.0

	<img width="1680" alt="Screen Shot 2020-06-12 at 12 51 08 PM" src="https://user-images.githubusercontent.com/1271364/84471781-0cf5a900-acac-11ea-9fb8-03b47523cb36.png">

	This PR

	<img width="1213" alt="Screen Shot 2020-06-12 at 12 49 33 PM" src="https://user-images.githubusercontent.com/1271364/84471807-1bdc5b80-acac-11ea-89e7-831826ce8a7d.png">

panzarino committed:
Merge branch 'develop' into issue-7642

panzarino committed:
Move more command progress to css

panzarino committed:
No progress bar for xhr

panzarino commented:
@jennifer-shehane I was able to fix a majority of the issues described here. 

As far as the debugging issue goes, the only way that I think this could be fixed is if there is some state that indicates that the tests have been opened in the debugger. Perhaps there could be an additional variable added to `appState` that indicates if the debugger is open, and that can be used to pause the animation.

The debugger gif you posted also revealed another small issue, that the progress bar would go back to full when complete, which has been fixed.

Fixed jitter effect:

![ezgif-6-e8673a2224fc](https://user-images.githubusercontent.com/7033952/84536611-374f6280-acbc-11ea-8af3-3b81d0368d44.gif)

Removed error where XHR requests would default to default command timeout:

![ezgif-6-620026d4b3bc](https://user-images.githubusercontent.com/7033952/84536829-a462f800-acbc-11ea-83fc-282c19040511.gif)

Performance improvements:

Current `develop` branch:

![Screen Shot 2020-06-12 at 3 07 38 PM](https://user-images.githubusercontent.com/7033952/84538024-decd9480-acbe-11ea-9c94-ea1040593ed4.png)

These changes:

![Screen Shot 2020-06-12 at 3 10 28 PM](https://user-images.githubusercontent.com/7033952/84538030-e12fee80-acbe-11ea-893a-7b438e41053f.png)

panzarino committed:
Merge branch 'develop' into issue-7642

panzarino committed:
Fix unit test

JessicaSachs commented:
I validated that the jitter is fixed.

Another issue: on failed tests the spinner fails to stop spinning.
<img width="447" alt="Screenshot 2020-06-12 15 29 20" src="https://user-images.githubusercontent.com/2801156/84539635-ce6ae900-acc1-11ea-9862-db06d53e5869.png">

I don't believe we need to pause the spinner when the debugger state is active.

panzarino commented:
@JessicaSachs That issue is actually present on `develop` and I think its been like that for a while now. I feel like it warrants its own issue and fixing it is outside the scope of this PR.

JessicaSachs commented on packages/reporter/src/commands/command.spec.tsx:
```diff
@@ -295,6 +297,12 @@ describe('<Command />', () => {
+    it('displays the timeout progress indicator', () => {
+      const component = shallow(<Command model={model({ state: 'pending' })} aliasesWithDuplicates={null} />)
+
+      expect(component.find(Progress).first()).to.exist
```
I do not believe this spec can fail because the `Progress` component is not being rendered conditionally... however, I'm not a React developer, so correct me if I'm wrong.

Do you understand why this spec was failing when you were looking for the DOM structure of `Progress` to assert the indicator was working?

JessicaSachs replied:
If we're going to use shallow mount, I would break this up into two specs:
1. Progress being rendered and passed the correct model
2. Progress having its own unit tests that are based on its internal structure + class

Alternatively, do a full mount (not shallow) and combine concerns 1 & 2

JessicaSachs replied:
A word of warning: testing timers can be particularly annoying at the unit/component level, so you could _instead_ choose to do an e2e spec to validate this, which may be easier to build and maintain 🙇‍♀️ 

Given the bugs we've been able to catch by manually testing, we should definitely add additional test coverage.

panzarino replied:
That is correct, but I'm not quite sure how to test it otherwise. Because of the way that the animation progress and time is calculated, its nearly impossible to mock that behavior unless somehow `Date.now()` can be stubbed. So, I've reverted to just ensuring that it gets rendered.

I believe it was failing due to the use of `shallow` which prevents assertions on child components. However, all of the "logic" around displaying the bar has now been moved to CSS, so again it's not conditional that the `div` gets rendered, so I figured just checking the component would be sufficient.

Ideally, there would be a way to ensure that the animations were being calculated correctly because that is the important thing to test here, but that requires stubbing `Date.now()`.

panzarino replied:
I agree that perhaps an `e2e` test would be better, also to ensure that the `log` model remains correct, however I am still unsure as to how to validate that the correct timing is calculated.

JessicaSachs committed:
Adding tests for progress indicator

panzarino committed:
Fix some types

panzarino committed:
Add e2e tests for progress bar

panzarino reviewed (commented):
Note: assertion timeout is not synced with command timeout, must fix before merge

JessicaSachs reviewed (dismissed):
Once the timeout issue you mentioned is fixed, this looks good to me

panzarino committed:
Remove timeout progress from assertion logs

panzarino commented:
I have removed the progress bars under each assertion so that there can just be one progress bar that is synchronized around the whole command.

@Bkucera and I decided that this progress bar should be displayed underneath the last assertion, but I was unable to find a way to accomplish this without a significant performance reduction. Therefore, the progress bar will be displayed underneath the top-level command as such:

![Screen Shot 2020-06-15 at 4 06 45 PM](https://user-images.githubusercontent.com/7033952/84701140-68c96780-af22-11ea-9851-df45d0bf2c5b.png)

This is probably not the ideal solution, but I am unable to think of any better viable solution.

panzarino committed:
Merge branch 'develop' into issue-7642

panzarino committed:
Merge branch 'develop' into issue-7642

JessicaSachs reviewed (approved):
Pulled it and ran it. Looks good to me.

jennifer-shehane reviewed (approved):
LGTM. I've rerun this firefox failing test a lot now and it keeps failing, so not sure there's something to investigate there.
